### PR TITLE
Improved interface for language variants

### PR DESCRIPTION
**Pull request by ryanachi**
This PR addresses https://github.com/goavki/apertium-html-tools/issues/82 (language variants picker). I've modified:

/assets/js/translator.js, to sort variants correctly
/assets/css/translation.css, to provide proper styling for variants
I've also added a new SQL file for variants in goavki/apertium-apy#78.

Please let me know if you have any comments!

### CODE REVIEW

File: assets/js/translator.js [hunk 1]
```diff
@@ -646,8 +653,41 @@ function populateTranslationList() {
...
+            if(srcLangs[i].indexOf('_') !== -1) {
+                variantsOnly.push(srcLangs[i]);
+            }
+            else {
+                langsOnly.push(srcLangs[i]);
+            }
+        }
+
+        srcLangs = langsOnly.sort(compareLangCodes);
+        for(i = 0; i < variantsOnly.length; i++) {
```
Comment (shardulc) [inline assets/js/translator.js]: While the double for-loop works, I think there's a more efficient way to do it:
1. sort `langsOnly`
2. sort `variantsOnly`
3. have `var langHead = 0` and `var variantHead = 0`
4. in a single for-loop, compare the lang at `langsOnly[langHead]` with the variant at `variantsOnly[variantHead]`, and depending on the comparison, either `langHead++` or `variantHead++`; simultaneously build the entire `srcLangs` list

(If that's not clear, ask on IRC :smiley:)

File: assets/js/translator.js [hunk 2]
```diff
@@ -1104,6 +1154,11 @@ function muteLanguages() {
             $(this).addClass('text-muted');
         }
     });
+    $.each($('#dstLanguages .languageVariant'), function () {
```
Comment (shardulc) [inline assets/js/translator.js]: Elements with the `languageVariant` class also have the `languageName` class, so the lines just above this already deal with them, right?

**Comment by shardulc** [comment]
Although it is displayed in the dropdown as just "Valencian", I think in the top bar it should be "Catalan Valencian":

![valencian][https://user-images.githubusercontent.com/10349624/34636584-a58483c2-f259-11e7-9b32-9a1ef08a4f01.png]

@sushain97 Do you agree, or is "Valencian" good?

**Comment by shardulc** [comment]
Another thing I'm not sure about: How does translating between variants work, e.g. Catalan -> Catalan Valencian? As far as I can see, it's not a 'pair' as such in Apertium, and apertium.org doesn't support it, so is it even possible?

**Comment by ryanachi** [comment]
@sushain97 would you like the top bar to display only the variant name, or both the language and variant names?

**Comment by sushain97** [comment]
> Although it is displayed in the dropdown as just "Valencian", I think in the top bar it should be "Catalan Valencian":

Consider for example the following three variants:

- por_BR
- cat_valencia
- eng_US

The language names DB could contain the "fully qualified" versions of each of these:

- Brazilian Portuguese
- Valencian
- American English

or the less fully qualified version:

- Brazilian
- Valencian
- American

Now our display will include the 'parent' language above the variants anyway so having the less qualified versions in the DB is fine UNTIL we try to put them into the recent languages bar when one gets selected. "American" doesn't seem like a good fit there. Consider that "I speak American" sounds somewhat awful. We could just append them together, "American English" is good. But then, "Valencian Catalan" sounds bad (I think?). We could put the variant in parentheses, e.g. "Catalan (Valencian)" or "Portuguese (Brazilian)". While that's less ideal, it does seem to work. Until you consider that we have to support all languages and parentheses are unlikely to be appropriate in all languages (the appending 'solution' probably has the same problem).

So, the solution seems to be either just using the fully qualified versions everywhere or having the DB store both (maybe semicolon delimited? e.g. the following and APy would split them:

- "Brazilian Portuguese;Brazilian"
- "Valencian;Valencian"
- "American;American English"

Storing both like this seems hacky and I think it would make the code pretty complicated. So, my thoughts are we go with the fully qualified version everywhere. Thoughts?

cc @jonorthwash @unhammer @ftyers thoughts on the right approach? have I missed any "cases"?

File: assets/css/translation.css [hunk 3]
```diff
@@ -6,6 +6,16 @@ html[dir='rtl'] .languageName {
     padding-right: .5em;
 }
 
+html[dir='ltr'] .languageVariant {
+	font-size: 12px;
```
Comment (sushain97) [inline assets/css/translation.css]: Indentation?

File: assets/css/translation.css [hunk 4]
```diff
@@ -6,6 +6,16 @@ html[dir='rtl'] .languageName {
...
+    padding-left: 2.5em;
+}
+
+html[dir='rtl'] .languageVariant {
+	font-size: 12px;
```
Comment (sushain97) [inline assets/css/translation.css]: As above

File: assets/js/translator.js [hunk 5]
```diff
@@ -567,8 +567,12 @@ function populateTranslationList() {
             if(numSrcLang < srcLangs.length) {
                 var langCode = srcLangs[j];
                 var langName = getLangByCode(langCode);
+                var langType = 'languageName';
```
Comment (sushain97) [inline assets/js/translator.js]: `langType` -> `langClasses`?

**Comment by ryanachi** [comment]
Adding the full names seems like a good idea; it definitely makes for a thorough solution.  I'll add the full names to the variants.sql file in APy.  Are there any other changes that we would like implemented?

**Comment by sushain97** [comment]
> Are there any other changes that we would like implemented?

Probably should work? :P See below.

image

**Comment by ryanachi** [comment]
That's odd; I'm getting this when I test. https://imgur.com/a/KP4zm

File: assets/js/translator.js [hunk 6]
```diff
@@ -646,8 +657,72 @@ function populateTranslationList() {
...
+
+        var langHead = 0;
+        var variantHead = 0;
+        var numLangs = srcLangs.length;
+        srcLangs = [];
+
+        langsOnly = langsOnly.sort(compareLangCodes);
+        variantsOnly = variantsOnly.sort(compareLangCodes);
+
+        while((langHead + variantHead) < numLangs) {
```
Comment (sushain97) [inline assets/js/translator.js]: I'm a little confused by this code. If the point of it is to get the order correct, I think `compareLangCodes` might already do that? Also, it should be extracted into a function since its used twice and is pretty long ;)

Comment (sushain97) [reply assets/js/translator.js]: (I didn't read the code super carefully so I could be misunderstand the intent)

**Comment by jonorthwash** [comment]
> Consider that "I speak American" sounds somewhat awful. 

[Not necessarily](http://www.pbs.org/speak/). But yeah, that's not what we want here.

**Comment by ryanachi** [comment]
@sushain97 I've incorporated the three suggestions that you made: https://395-28576409-gh.circle-artifacts.com/0/home/circleci/project/build/index.eng.html?dir=slv-bos#translation

File: assets/js/translator.js [hunk 7]
```diff
@@ -647,20 +682,7 @@ function populateTranslationList() {
...
-                return compareLangCodes(a, b);
-            }
-            else if(aPossible && !bPossible) {
-                return -1;
-            }
-            else {
-                return 1;
-            }
-        });
+        dstLangs = dstLangs.sort(compareLangCodes);
```
Comment (sushain97) [inline assets/js/translator.js]: Why was this deleted? The extra sorting puts available languages first. It might require some modification for variants properly displaying though.

File: assets/js/translator.js [hunk 8]
```diff
@@ -635,10 +670,10 @@ function populateTranslationList() {
             if(aVariant[1] && bVariant[1] && aVariant[0] === bVariant[0]) {
                 return directCompare;
             }
-            else if(aVariant && a.startsWith(b)) {
+            else if(aVariant[1] && aVariant[0] === b) {
```
Comment (sushain97) [inline assets/js/translator.js]: Good catch.

File: assets/js/translator.js [hunk 9]
```diff
@@ -600,10 +624,22 @@ function populateTranslationList() {
...
+                indent += '&nbsp;&nbsp;&nbsp;&nbsp;';
+            }
+        }
+        $('#srcLangSelect').append($('<option></option>').prop('value', this).html(indent + getLangByCode(this)));
     });
     $.each(dstLangs, function () {
-        $('#dstLangSelect').append($('<option></option>').prop('value', this).text(getLangByCode(this)));
+        var indent = '';
+        if(typeof this !== 'undefined') {
+            if(this.indexOf('_') !== -1) {
```
Comment (sushain97) [inline assets/js/translator.js]: Can these be combined into a single if with a conjunction?

**Comment by ryanachi** [comment]
Done! https://397-28576409-gh.circle-artifacts.com/0/home/circleci/project/build/index.eng.html?dir=slv-bos#translation

**Comment by sushain97** [comment]
Now:

image

Before:

image

It doesn't look like we're getting the same behavior where available languages are sorted first.

**Comment by sushain97** [comment]
sad:

image

**Comment by ryanachi** [comment]
@sushain97 I've revised the sorting function to correctly display the language variants.

File: assets/js/translator.js [hunk 10]
```diff
@@ -661,6 +699,16 @@ function populateTranslationList() {
                 return 1;
             }
         });
+
+        for(i = 0; i < variantsOnly.length; i++) {
```
Comment (sushain97) [inline assets/js/translator.js]: Just to confirm - this works when a variant is possible but the parent language isn't and vice versa, right? Could you provide screenshots for each case? Also, this nested for loop seems a bit expensive?

Comment (ryanachi) [reply assets/js/translator.js]: Yes, that's correct—I've attached a screenshot of a case when the parent language is available but the variant language isn't.  As for the converse scenario, apertium.org doesn't have a pair where the variant language is available but the parent language isn't.  However, it should work as well, as the 'for' loop inserts each variant language immediately after its respective parent language, no matter what the position is—this can be seen in the screenshot.

I agree, the 'for' loop is a bit inefficient; however, I couldn't find a better way to sort the list of languages so that each variant appears immediately after its parent language.  I'd be happy to implement another solution if we can find one. :)

**Comment by sushain97** [comment]
What if you sorted the same way (built in sort) but considered a variant
available if it's actually available, it's parent is available or another
variant with the same parent is available. And for a non-variant, available
if it's actually available or any of it's variants is available. Building a
hash from language to its variants will make this effectively linear,
right? Splicing is also pretty expensive when you do it over and over again.

**Comment by sushain97** [comment]
Also, try it out with beta.apertium.org/apy, it's probably easier to find a
variant that's possible with a parent that isn't then.

On Jan 9, 2018 12:47 PM, "Sushain Cherivirala" <sushain@skc.name> wrote:

What if you sorted the same way (built in sort) but considered a variant
available if it's actually available, it's parent is available or another
variant with the same parent is available. And for a non-variant, available
if it's actually available or any of it's variants is available. Building a
hash from language to its variants will make this effectively linear,
right? Splicing is also pretty expensive when you do it over and over again.

File: assets/js/translator.js [hunk 11]
```diff
@@ -648,9 +675,22 @@ function populateTranslationList() {
 
         srcLangs = srcLangs.sort(compareLangCodes);
         dstLangs = dstLangs.sort(function (a, b) {
-            var aPossible = pairs[curSrcLang] && pairs[curSrcLang].indexOf(a) !== -1;
-            var bPossible = pairs[curSrcLang] && pairs[curSrcLang].indexOf(b) !== -1;
-
+            function determineIfPossible(a) {
```
Comment (sushain97) [inline assets/js/translator.js]: Use `isPossible(x)` instead so we don't shadow `a`.

File: assets/js/translator.js [hunk 12]
```diff
@@ -648,9 +675,22 @@ function populateTranslationList() {
...
+                var aVariant = a.split('_');
+                if(pairs[curSrcLang] && pairs[curSrcLang].indexOf(a) !== -1) {
```
Comment (sushain97) [inline assets/js/translator.js]: Also this should just be a huge return with each of these things as parenthesized lines in a conjunction :)

File: assets/js/translator.js [hunk 13]
```diff
@@ -648,9 +675,22 @@ function populateTranslationList() {
...
+                    return true;
+                }
+                else if(pairs[curSrcLang] && pairs[curSrcLang].indexOf(aVariant[0]) !== -1) {
+                    return true;
+                }
+                else if(pairs[curSrcLang] && pairs[curSrcLang].join().match('/' + aVariant[0] + '/g')) {
```
Comment (sushain97) [inline assets/js/translator.js]: This is really hacky..

Comment (sushain97) [reply assets/js/translator.js]: Maybe `.filter`?
